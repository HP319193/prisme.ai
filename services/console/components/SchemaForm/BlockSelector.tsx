import {
  InfoBubble,
  Schema,
  SchemaForm,
  schemaFormUtils,
  Select,
  StretchContent,
  useSchemaForm,
} from '@prisme.ai/design-system';
import { FieldComponent } from '@prisme.ai/design-system/lib/Components/SchemaForm/context';
import { useField } from 'react-final-form';
import { Tooltip } from 'antd';
import FieldContainerWithRaw from '../FieldContainerWithRaw';
import useBlocks from '../PageBuilder/useBlocks';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useTranslation } from 'next-i18next';
import useLocalizedText from '../../utils/useLocalizedText';
import getEditSchema from '../PageBuilder/Panel/EditSchema/getEditSchema';
import { RightOutlined } from '@ant-design/icons';

export const BlockSelector: FieldComponent = (props) => {
  const { t } = useTranslation('workspaces');
  const { localize, localizeSchemaForm } = useLocalizedText();
  const field = useField(props.name);
  const hasError = schemaFormUtils.getError(field.meta);
  const { variants: blocks } = useBlocks();
  const [schema, setSchema] = useState<Schema | null>(null);
  const [values, setValues] = useState(field.input.value || {});
  const { utils, locales, components } = useSchemaForm();
  const [visible, setVisible] = useState(true);

  const fetchSchema = useCallback(
    async (name: string) => {
      await setSchema(null);
      const block = blocks.find(({ slug }) => slug === name);
      if (!block) return null;
      if (block.builtIn) {
        const schema = getEditSchema(block.slug);
        setSchema(schema && localizeSchemaForm(schema));
      }
    },
    [blocks, localizeSchemaForm]
  );

  useEffect(() => {
    fetchSchema(field.input.value?.name);
  }, [fetchSchema, field.input.value?.name]);

  const prevValues = useRef(values);
  useEffect(() => {
    if (prevValues.current === values) return;
    const newValue = {
      ...(field.input.value || {}),
      ...values,
    };
    if (JSON.stringify(field.input.value) === JSON.stringify(newValue)) return;
    field.input.onChange(newValue);
  }, [field.input, values]);

  const selectOptions = useMemo(
    () =>
      blocks.map(({ slug, name, description, photo }) => ({
        label: (
          <Tooltip
            title={
              localize(description) ? (
                <>
                  {localize(description)}
                  {photo && (
                    // eslint-disable-next-line @next/next/no-img-element
                    <img
                      src={photo}
                      alt={localize(name)}
                      className="bg-white rounded"
                    />
                  )}
                </>
              ) : (
                ''
              )
            }
          >
            <div className="flex flex-1">{localize(name)}</div>
          </Tooltip>
        ),
        value: slug,
      })),
    [blocks, localize]
  );

  const filterOption = useCallback(
    (input: string, options: any) => {
      const block = blocks.find(({ slug }) => slug === options.value);
      if (!block) return false;
      const search = `${block.slug} ${localize(block.name)} ${localize(
        block.description
      )}`.toLowerCase();
      return search.includes(input);
    },
    [blocks, localize]
  );

  return (
    <FieldContainerWithRaw
      {...props}
      className="pr-form-block-selector pr-form-object"
    >
      <button
        type="button"
        className={`pr-form-object__label pr-form-label ${
          visible ? 'pr-form-object__label--visible' : ''
        }`}
        onClick={() => setVisible(!visible)}
      >
        <RightOutlined className="pr-form-object__label-icon" />

        <span>
          {props.schema.title || schemaFormUtils.getLabel(field.input.name)}
        </span>

        <InfoBubble
          className="pr-form-object__description"
          text={props.schema.description}
        />
      </button>
      <StretchContent visible={visible}>
        <div className="pr-form-object__properties">
          <Tooltip title={hasError} overlayClassName="pr-form-error">
            <div className="flex-col m-[1rem]">
              <div className="flex flex-col">
                <label
                  htmlFor={`${field.input.name}.name`}
                  className="pr-form-block-selector__label pr-form-label"
                >
                  {t('form.blockSelector.name.label')}
                </label>
                <Select
                  id={`${field.input.name}.name`}
                  selectOptions={selectOptions}
                  value={field.input.value?.name}
                  onChange={(name) => {
                    setValues({ name });
                  }}
                  showSearch
                  filterOption={filterOption}
                />
              </div>
              {schema && (
                <div className="!-mx-[1rem]">
                  <SchemaForm
                    schema={schema}
                    locales={locales}
                    buttons={[]}
                    initialValues={values}
                    utils={utils}
                    components={components}
                    onChange={(v) => {
                      setValues((prev: any) => {
                        const newValue = {
                          ...prev,
                          ...v,
                        };
                        if (JSON.stringify(prev) === JSON.stringify(newValue))
                          return prev;
                        return newValue;
                      });
                    }}
                  />
                </div>
              )}
            </div>
          </Tooltip>
        </div>
      </StretchContent>
    </FieldContainerWithRaw>
  );
};

export default BlockSelector;
