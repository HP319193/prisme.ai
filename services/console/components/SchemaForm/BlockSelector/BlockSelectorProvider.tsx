import { BlockComponent, loadModule } from '@prisme.ai/blocks';
import { Schema } from '@prisme.ai/design-system';
import { createContext, FC, useCallback, useEffect, useState } from 'react';
import { useContext } from '../../../utils/useContext';
import useLocalizedText from '../../../utils/useLocalizedText';
import getEditSchema from '../../PageBuilder/Panel/EditSchema/getEditSchema';
import useBlocks, { BlockInCatalog } from '../../PageBuilder/useBlocks';

interface BlockSelectorContext {
  blocks: BlockInCatalog[];
  selectBlock: (selected: string | null) => void;
  selectedBlock: BlockInCatalog;
  schema: Schema | null;
}

export const blockSelectorContext = createContext<
  BlockSelectorContext | undefined
>(undefined);

export const useBlockSelector = () =>
  useContext<BlockSelectorContext>(blockSelectorContext);

export const BlockSelectorProvider: FC = ({ children }) => {
  const { variants: blocks } = useBlocks();
  const [selectedBlock, setSelectedBlock] = useState<BlockInCatalog | null>(
    null
  );
  const [schema, setSchema] = useState<Schema | null>(null);
  const { localizeSchemaForm } = useLocalizedText();

  const fetchSchema = useCallback(
    async (block: BlockInCatalog | null) => {
      // This await goal is to force a re-render
      await setSchema(null);
      if (!block) return null;
      if (block.builtIn) {
        const schema = getEditSchema(block.slug);
        setSchema(schema && localizeSchemaForm(schema));
        return;
      }
      if (block.url) {
        const module = await loadModule<BlockComponent>(block.url);
        if (module && module.schema) {
          setSchema(module.schema);
        }
      }
    },
    [localizeSchemaForm]
  );

  const selectBlock: BlockSelectorContext['selectBlock'] = useCallback(
    (blockSlug) => {
      const block = blocks.find(({ slug }) => slug === blockSlug);
      setSelectedBlock(block || null);
      fetchSchema(block || null);
    },
    [blocks, fetchSchema]
  );

  return (
    <blockSelectorContext.Provider
      value={{
        blocks,
        selectBlock,
        schema,
      }}
    >
      {children}
    </blockSelectorContext.Provider>
  );
};
