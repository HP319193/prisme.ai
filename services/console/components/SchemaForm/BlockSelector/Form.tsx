import { SchemaForm, useSchemaForm } from '@prisme.ai/design-system';
import { useBlockSelector } from './BlockSelectorProvider';

interface SchemaFormProps {
  values: Record<string, any>;
  onChange: (values: SchemaFormProps['values']) => void;
}

export const Form = ({ values, onChange }: SchemaFormProps) => {
  const { schema } = useBlockSelector();
  const { utils, locales, components } = useSchemaForm();

  if (!schema) return null;

  return (
    <SchemaForm
      schema={schema}
      locales={locales}
      buttons={[]}
      initialValues={values}
      utils={utils}
      components={components}
      onChange={onChange}
    />
  );
};

export default Form;
