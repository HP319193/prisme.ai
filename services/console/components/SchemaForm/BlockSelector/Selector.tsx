import { Select, Tooltip } from '@prisme.ai/design-system';
import { useCallback, useMemo } from 'react';
import useLocalizedText from '../../../utils/useLocalizedText';
import { useBlockSelector } from './BlockSelectorProvider';

export interface SchemaSelectorProps {
  id?: string;
  value: string;
  onChange: (v: string) => void;
}

export const Selector = ({ id = '', value, onChange }: SchemaSelectorProps) => {
  const { blocks } = useBlockSelector();
  const { localize } = useLocalizedText();
  const selectOptions = useMemo(
    () =>
      blocks.map(({ slug, name, description, photo }) => ({
        label: (
          <Tooltip
            title={
              localize(description) ? (
                <>
                  {localize(description)}
                  {photo && (
                    // eslint-disable-next-line @next/next/no-img-element
                    <img
                      src={photo}
                      alt={localize(name)}
                      className="bg-white rounded"
                    />
                  )}
                </>
              ) : (
                ''
              )
            }
          >
            <div className="flex flex-1">{localize(name)}</div>
          </Tooltip>
        ),
        value: slug,
      })),
    [blocks, localize]
  );

  const filterOption = useCallback(
    (input: string, options: any) => {
      const block = blocks.find(({ slug }) => slug === options.value);
      if (!block) return false;
      const search = `${block.slug} ${localize(block.name)} ${localize(
        block.description
      )}`.toLowerCase();
      return search.includes(input);
    },
    [blocks, localize]
  );

  return (
    <Select
      id={id}
      selectOptions={selectOptions}
      value={value}
      onChange={onChange}
      showSearch
      filterOption={filterOption}
    />
  );
};

export default Selector;
