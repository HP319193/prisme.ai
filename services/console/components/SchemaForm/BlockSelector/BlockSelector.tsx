import {
  InfoBubble,
  schemaFormUtils,
  StretchContent,
} from '@prisme.ai/design-system';
import { FieldComponent } from '@prisme.ai/design-system/lib/Components/SchemaForm/context';
import { useField } from 'react-final-form';
import { Tooltip } from 'antd';
import FieldContainerWithRaw from '../../FieldContainerWithRaw';
import { useEffect, useRef, useState } from 'react';
import { useTranslation } from 'next-i18next';
import { RightOutlined } from '@ant-design/icons';
import Selector from './Selector';
import {
  BlockSelectorProvider,
  useBlockSelector,
} from './BlockSelectorProvider';
import Form from './Form';

export const BlockSelector: FieldComponent = (props) => {
  const { selectBlock, schema } = useBlockSelector();
  const { t } = useTranslation('workspaces');
  const field = useField(props.name);
  const hasError = schemaFormUtils.getError(field.meta);
  const [values, setValues] = useState(field.input.value || {});
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    selectBlock(field.input.value?.slug);
  }, [field.input.value?.slug, selectBlock]);

  const prevValues = useRef(values);
  useEffect(() => {
    if (prevValues.current === values) return;
    const newValue = {
      ...(field.input.value || {}),
      ...values,
    };
    if (JSON.stringify(field.input.value) === JSON.stringify(newValue)) return;
    field.input.onChange(newValue);
  }, [field.input, values]);

  return (
    <FieldContainerWithRaw
      {...props}
      className="pr-form-block-selector pr-form-object"
    >
      <button
        type="button"
        className={`pr-form-object__label pr-form-label ${
          visible ? 'pr-form-object__label--visible' : ''
        }`}
        onClick={() => setVisible(!visible)}
      >
        <RightOutlined className="pr-form-object__label-icon" />

        <span>
          {props.schema.title || schemaFormUtils.getLabel(field.input.name)}
        </span>

        <InfoBubble
          className="pr-form-object__description"
          text={props.schema.description}
        />
      </button>
      <StretchContent visible={visible}>
        <div className="pr-form-object__properties">
          <Tooltip title={hasError} overlayClassName="pr-form-error">
            <div className="flex-col m-[1rem]">
              <div className="flex flex-col">
                <label
                  htmlFor={`${field.input.name}.name`}
                  className="pr-form-block-selector__label pr-form-label"
                >
                  {t('form.blockSelector.name.label')}
                </label>
                <Selector
                  id={`${field.input.name}.slug`}
                  value={field.input.value?.slug}
                  onChange={(slug) => {
                    setValues({ slug });
                  }}
                />
              </div>
              {schema && (
                <div className="!-mx-[1rem]">
                  <Form
                    values={values}
                    onChange={(v) => {
                      setValues((prev: any) => {
                        const newValue = {
                          ...prev,
                          ...v,
                        };
                        if (JSON.stringify(prev) === JSON.stringify(newValue))
                          return prev;
                        return newValue;
                      });
                    }}
                  />
                </div>
              )}
            </div>
          </Tooltip>
        </div>
      </StretchContent>
    </FieldContainerWithRaw>
  );
};

export const LinkedBlockSelector: FieldComponent = (props) => (
  <BlockSelectorProvider>
    <BlockSelector {...props} />
  </BlockSelectorProvider>
);

export default LinkedBlockSelector;
