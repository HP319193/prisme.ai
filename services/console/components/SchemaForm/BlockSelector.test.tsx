import renderer from 'react-test-renderer';
import BlockSelector from './BlockSelector';

jest.mock('react-final-form', () => {
  const mock = {
    input: {
      value: '',
      onChange: jest.fn(),
    },
    meta: {},
  };
  return {
    useField: () => mock,
  };
});

jest.mock('../PageBuilder/useBlocks', () => {
  const mock = { variants: [] };
  return () => mock;
});

jest.mock('../PageBuilder/Panel/EditSchema/getEditSchema', () => {
  return () => ({});
});

it('should render', () => {
  const root = renderer.create(<BlockSelector name="foo" schema={{}} />);
  expect(root.toJSON()).toMatchSnapshot();
});
