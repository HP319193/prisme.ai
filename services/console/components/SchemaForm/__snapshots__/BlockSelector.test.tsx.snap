// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="pr-form-field pr-form-block-selector pr-form-object"
>
  <button
    className="pr-form-raw absolute top-0 right-0 mt-[0.35rem] flex flex-1 flex-row z-[1] text-[12px] items-center"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  />
  <button
    className="pr-form-object__label pr-form-label "
    onClick={[Function]}
    type="button"
  >
    <span />
  </button>
  <div
    className="pr-form-object__properties"
  >
    <div
      className="flex-col m-[1rem]"
    >
      <div
        className="flex flex-col"
      >
        <label
          className="pr-form-block-selector__label pr-form-label"
          htmlFor="undefined.name"
        >
          form.blockSelector.name.label
        </label>
      </div>
    </div>
  </div>
</div>
`;
